fontra2fontir/src/source.rs

```diff
@@ -1,22 +1,146 @@
-use std::path::PathBuf;
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::{BufRead, BufReader},
+    path::PathBuf,
+};
 
-use fontir::source::Source;
+use fontdrasil::types::GlyphName;
+use fontir::{
+    error::Error,
+    source::{Input, Source},
+    stateset::StateSet,
+};
+use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
 
 pub struct FontraIrSource {
-    _fontra_dir: PathBuf,
+    fontdata_file: PathBuf,
+    glyphinfo_file: PathBuf,
+    glyph_dir: PathBuf,
+    glyph_info: HashMap<GlyphName, (PathBuf, Option<u32>)>,
 }
 
 impl FontraIrSource {
-    pub fn new(fontra_dir: PathBuf) -> Self {
-        FontraIrSource {
-            _fontra_dir: fontra_dir,
+    pub fn new(fontra_dir: PathBuf) -> Result<Self, Error> {
+        let fontdata_file = fontra_dir.join("font-data.json");
+        if !fontdata_file.is_file() {
+            return Err(Error::FileExpected(fontdata_file));
         }
+        let glyphinfo_file = fontra_dir.join("glyph-info.csv");
+        if !glyphinfo_file.is_file() {
+            return Err(Error::FileExpected(glyphinfo_file));
+        }
+        let glyph_dir = fontra_dir.join("glyphs");
+        if !glyph_dir.is_dir() {
+            return Err(Error::DirectoryExpected(glyph_dir));
+        }
+        Ok(FontraIrSource {
+            fontdata_file,
+            glyphinfo_file,
+            glyph_dir,
+            glyph_info: Default::default(),
+        })
+    }
+
+    fn glyph_file(&self, glyph: GlyphName) -> PathBuf {
+        // TODO: this probably over-encodes. Should be sufficient to get started.
+        let filename = utf8_percent_encode(glyph.as_str(), NON_ALPHANUMERIC);
```

Comment:
I think fontra's glyph->filename algorithm is implemented here:
https://github.com/googlefonts/fontra/blob/main/src/fontra/backends/filenames.py